include/instantiate.h

```diff
@@ -232,7 +232,118 @@ namespace quda
 #endif
     } else if (field.Precision() == QUDA_QUARTER_PRECISION) {
 #if QUDA_PRECISION & 1
-      instantiate<Apply, char>(field, args...);
+      instantiate<Apply, int8_t>(field, args...);
+#else
+      errorQuda("QUDA_PRECISION=%d does not enable quarter precision", QUDA_PRECISION);
+#endif
+    } else {
+      errorQuda("Unsupported precision %d\n", field.Precision());
+    }
+  }
+
+  /**
+     @brief The instantiatePrecision function is used to instantiate
+     the precision
+     @param[in] field LatticeField we wish to instantiate
+     @param[in,out] args Any additional arguments required for the
+     computation at hand
+  */
+  template <template <typename> class Apply, typename F, typename... Args>
+  constexpr void instantiatePrecision(F &field, Args &&... args)
+  {
+    if (field.Precision() == QUDA_DOUBLE_PRECISION) {
+      // always instantiate double precision
+      Apply<double>(field, args...);
+    } else if (field.Precision() == QUDA_SINGLE_PRECISION) {
+#if QUDA_PRECISION & 4
+      Apply<float>(field, args...);
+#else
+      errorQuda("QUDA_PRECISION=%d does not enable single precision", QUDA_PRECISION);
+#endif
+    } else if (field.Precision() == QUDA_HALF_PRECISION) {
+#if QUDA_PRECISION & 2
+      Apply<short>(field, args...);
+#else
+      errorQuda("QUDA_PRECISION=%d does not enable half precision", QUDA_PRECISION);
+#endif
+    } else if (field.Precision() == QUDA_QUARTER_PRECISION) {
+#if QUDA_PRECISION & 1
+      Apply<int8_t>(field, args...);
+#else
+      errorQuda("QUDA_PRECISION=%d does not enable quarter precision", QUDA_PRECISION);
+#endif
+    } else {
+      errorQuda("Unsupported precision %d\n", field.Precision());
+    }
+  }
+
+  /**
+     @brief The instantiatePrecision function is used to instantiate
```

Comment:
Why do we need two `instantiatePrecision` functions (different template signatures noted)? If having two is unavoidable, is there a better naming option than `instantiatePrecision` and `instantiatePrecision2`?

Reply:
Cleaning this up is easy with C++17 😜   This solution was expedient in the meantime....  Open to naming suggestions.  This is really just a placeholder.

Reply:
Fair, just please make the use for this version vs the other one clearer from the doxygen

Reply:
Will do.